pkg/search/proxy/store/multi_cluster_cache.go

```diff
@@ -334,15 +334,34 @@ func (c *MultiClusterCache) Watch(ctx context.Context, gvr schema.GroupVersionRe
 		if cache == nil {
 			continue
 		}
-		w, err := cache.Watch(ctx, options)
-		if err != nil {
+
+		// The following logic adds a 30-second timeout to prevent watch requests to member clusters from hanging,
+		// which could cause the client watch to hang.
+		watchChan := make(chan watch.Interface, 1)
+		errChan := make(chan error, 1)
+
+		go func() {
+			w, err := cache.Watch(ctx, options)
+			if err != nil {
+				errChan <- fmt.Errorf("failed to start watch for resource %v in cluster %q: %v", gvr.String(), cluster, err)
+				return
+			}
+			watchChan <- w
+		}()
+
+		select {
+		case w := <-watchChan:
+			mux.AddSource(w, func(e watch.Event) {
+				setObjectResourceVersionFunc(cluster, e.Object)
+				addCacheSourceAnnotation(e.Object, cluster)
+			})
+		case err := <-errChan:
+			// If the watch request fails, return the error, and the client will retry.
 			return nil, err
+		case <-time.After(30 * time.Second):
+			// If the watch request times out, return an error, and the client will retry.
+			return nil, fmt.Errorf("timeout waiting for watch for resource %v in cluster %q", gvr.String(), cluster)
```

Comment:
@xigang Hi, if a watch request is hanging and causes a timeout, will the hanging watch request continue to exist in the subprocess?

Reply:
@zhzhuang-zju Yes, there is this issue. When a watch request times out, the goroutine needs to be terminated.

Reply:
Good point! Then that case we have to cancel the context passed to cache.Watch().

Reply:
Another question:
Before starting the `Watch`, we tried to get the cache of that cluster, I'm curious why this cache still exists even after the cluster is gone. Do we have a chance to clean the cache?

https://github.com/karmada-io/karmada/blob/e7b6513db0e44090a9d11fcff15b1fec097b3b2a/pkg/search/proxy/store/multi_cluster_cache.go#L333-L336

Reply:
> Another question: Before starting the `Watch`, we tried to get the cache of that cluster, I'm curious why this cache still exists even after the cluster is gone. Do we have a chance to clean the cache?
> 
> https://github.com/karmada-io/karmada/blob/e7b6513db0e44090a9d11fcff15b1fec097b3b2a/pkg/search/proxy/store/multi_cluster_cache.go#L333-L336

@RainbowMango When the member cluster goes offline but the Cluster resources in the control plane are not deleted, it can prevent the offline clusters in the ResourceRegistry from being removed, resulting in the resource cache being retained for a short time.

Reply:
> @xigang Hi, if a watch request is hanging and causes a timeout, will the hanging watch request continue to exist in the subprocess?

@zhzhuang-zju Fixed, please take a look.